File: gateway/auth/ts/src/tlog/sth.ts
/**
 * Signed tree head: signing input, signing (for tests and any TS log
 * producer), and verification against the gateway's `log_pubkey`.
 */

import { bytesToHex, hexToBytes } from "../encoding.js";
import { signingBytes } from "../jcs.js";
import { ecdsaSign, ecdsaVerify } from "../sigs.js";
import type { SignedTreeHead, SignedTreeHeadUnsigned } from "./types.js";

/**
 * The exact bytes the log key signs: `JCS({tree_size, root_hash,
 * signed_at})`. Built from the three named fields only, so an object
 * carrying extra keys still yields the producer's bytes.
 */
export function sthSigningBytes(sth: SignedTreeHeadUnsigned): Uint8Array {
  return signingBytes(
    { tree_size: sth.tree_size, root_hash: sth.root_hash, signed_at: sth.signed_at },
    "sig",
  );
}

/** Sign a tree head with a 32-byte secp256k1 private key. */
export function signSth(unsigned: SignedTreeHeadUnsigned, logPrivKey: Uint8Array): SignedTreeHead {
  return {
    tree_size: unsigned.tree_size,
    root_hash: unsigned.root_hash,
    signed_at: unsigned.signed_at,
    sig: bytesToHex(ecdsaSign(logPrivKey, sthSigningBytes(unsigned))),
  };
}

/**
 * Verify `sth.sig` with the compressed-hex `log_pubkey` served next
 * to it. The log key is per gateway boot and attests only "this
 * gateway served this head"; a witness must still recompute the root
 * from the leaves before countersigning.
 */
export function verifySth(sth: SignedTreeHead, logPubkeyHex: string): boolean {
  try {
    return ecdsaVerify(hexToBytes(logPubkeyHex), sthSigningBytes(sth), hexToBytes(sth.sig));
  } catch {
    return false;
  }
}

File: gateway/internal/tlog/sth.go
package tlog

import (
	"fmt"
	"time"

	macaroon "github.com/stakwork/stakgraph/gateway/auth/go"
)

// STH is a signed tree head. Sig is the log key's ECDSA signature over
// SHA256(JCS(sth \ sig)), the same construction as the macaroon
// layers, so gatekey verifies it with the primitives it already has.
//
// The empty tree is a valid, signed head: tree_size 0 and the RFC 9162
// empty root.
type STH struct {
	TreeSize uint64 `json:"tree_size"`
	RootHash string `json:"root_hash"`
	SignedAt string `json:"signed_at"`
	Sig      string `json:"sig"`
}

// SigningBytes returns the exact bytes the log key signs for sth:
// JCS of the head with the sig field removed.
func SigningBytes(sth STH) ([]byte, error) {
	return macaroon.JCSStripField(sth, "sig")
}

// VerifySTH checks sth.Sig against a compressed-hex log pubkey. This
// is the Go twin of gatekey's verifySth, for tests and any Go witness.
func VerifySTH(sth STH, logPubkeyHex string) bool {
	msg, err := SigningBytes(sth)
	if err != nil {
		return false
	}
	pub, err := macaroon.HexToBytes(logPubkeyHex)
	if err != nil {
		return false
	}
	sig, err := macaroon.HexToBytes(sth.Sig)
	if err != nil {
		return false
	}
	return macaroon.EcdsaSecp256k1Verify(pub, msg, sig)
}

func signSTH(key *logKey, size uint64, root Hash, at time.Time) (STH, error) {
	sth := STH{
		TreeSize: size,
		RootHash: macaroon.BytesToHex(root[:]),
		SignedAt: formatTime(at),
	}
	msg, err := SigningBytes(sth)
	if err != nil {
		return STH{}, fmt.Errorf("tlog: sth signing bytes: %w", err)
	}
	sig, err := key.sign(msg)
	if err != nil {
		return STH{}, fmt.Errorf("tlog: sign sth: %w", err)
	}
	sth.Sig = macaroon.BytesToHex(sig)
	return sth, nil
}

// formatTime renders an RFC 3339 UTC timestamp with millisecond
// precision, the format used for leaf.ts and sth.signed_at.
func formatTime(t time.Time) string {
	return t.UTC().Format("2006-01-02T15:04:05.000Z07:00")
}
